src/lib/jwt.js

```diff
@@ -1,70 +1,144 @@
-const DEFAULT_ENCRYPTION_TYPE = 'AES' // Can be one of [ 'AES', false ]
+// Set default algorithm to use for signing
+const DEFAULT_SIGNATURE_ALGORITHM = 'HS512'
+const DEFAULT_SIGNATURE_ALGORITHMS = [ DEFAULT_SIGNATURE_ALGORITHM ]
+
+// Set default algorithm to use for encryption
+// YOU SHOULD NOT USE PBES2-HS512+A256KW IN PRODUCTION IN THIS WAY
+// (i.e. with a shared secret where a token can be trivially re-generated)
+// JSON Web Algorithms (JWA) RFC 7518 recommends ECDH-ES+A256KW
+const DEFAULT_ENCRYPTION_ALGORITHM = 'PBES2-HS512+A256KW'
+const DEFAULT_ENCRYPTION_ALGORITHMS = [ DEFAULT_ENCRYPTION_ALGORITHM ]
+
+// Generate warning when keys not set, but don't want to flood the logs
+let SIGNING_KEY_WARNING = false
+let ENCRYPTION_KEY_WARNING = false
 
 const encode = async ({
-  secret,
-  key = secret,
   token = {},
   maxAge,
-  encryption = DEFAULT_ENCRYPTION_TYPE
-}) => {
-  // If maxAge is set remove any existing created/expiry dates and replace them
-  if (maxAge) {
-    if (token.iat) { delete token.iat }
-    if (token.exp) { delete token.exp }
+  secret,
+  encryptionPrivateKey,
+  encryptionPublicKey,
+  signingPrivateKey,
+  signingPublicKey,
+  issuer,
+  encryption = true
+} = {}) => {
+
+  // Signing Key
+  let signingKeyValue = signingPublicKey || signingPrivateKey
+  if (!signingKeyValue) {
+    if (!SIGNING_KEY_WARNING) {
+      logger.warn('JWT_ENCODE_AUTO_GENERATED_SIGNING_KEY')
+      SIGNING_KEY_WARNING = true
+    }
+    const autoGeneratedSigningSecret = createHash('sha256').update(secret).digest('hex')
+    signingKeyValue = `{"kty":"oct","kid":"nextauth-auto-generated-signing-key","alg":"${DEFAULT_SIGNATURE_ALGORITHM}","k":"${autoGeneratedSigningSecret}"}`
+  }
+  const signingKeyInput = JSON.parse(signingKeyValue)
+  const signingKey = await jose.JWK.asKey(signingKeyInput)
+
+  const options = {
+    expiresIn: `${maxAge}s`,
+    kid: false,
   }
 
-  const signedToken = jwt.sign(token, secret, { algorithm: SIGNING_ALGORITHM, expiresIn: maxAge })
+  if (issuer) { options.issuer = issuer }
 
-  switch (encryption) {
-    case false:
-      return signedToken
-    case 'AES':
-      return CryptoJS.AES.encrypt(signedToken, key).toString()
-    default:
-      throw new Error('Unsupported value for `encryption` passed to JWT encode()', encryption)
+  const signedToken = jose.JWT.sign(token, signingKey, options)
+
+  if (encryption) {
+    // Encryption Key
+    let encryptionKeyValue = encryptionPublicKey || encryptionPrivateKey
+    if (!encryptionKeyValue) {
+      if (!ENCRYPTION_KEY_WARNING) {
+        logger.warn('JWT_ENCODE_AUTO_GENERATED_ENCRYPTION_KEY')
+        ENCRYPTION_KEY_WARNING = true
+      }
+      const autoGeneratedEncryptionSecret = createHash('sha256').update(secret).digest('hex')
+      encryptionKeyValue = `{"kty":"oct","kid":"nextauth-auto-generated-encryption-key","alg":"${DEFAULT_ENCRYPTION_ALGORITHM}","k":"${autoGeneratedEncryptionSecret}"}`
```

Comment:
> `"k":"${autoGeneratedEncryptionSecret}"}`

This is not what `k` should be :) `k` are supposed to be base64url encoded octets.

So, e.g. When you have a 32 byte buffer to be used for `A256GCM` direct encryption just instantiate your key like so

```js
const key = jose.JWK.asKey(buffer, { alg: 'A256GCM', use: 'enc', kid: 'nextauth-auto-generated-encryption-key' }
```

Reply:
Thanks! Yes this was terrible but 'works' as the hash didn't contain any characters that are not valid for Base64 encoding (and didn't get stored anywhere except, in memory so nothing mangled it). HKDF is a MUCH better solution and I will definitely make that the default method for key generation instead of this hackery.